### PR TITLE
SLEP015: Feature Names Propagation

### PR DESCRIPTION
**Pull request by thomasjpfan**
This SLEP details how a `feature_names_in_` attribute and a `get_feature_names_out` method can be used together to obtain feature name propagation. I see there are two main goals for having this feature:

1. Inspect the feature names inputted into the last step of a pipeline. This is address as the motivating problem in the SLEP.
2. Be able to use panda's dataframes or xarray's dataarray as input for estimators in a pipeline. This allows for estimators to do something special based on the name of a feature. This is addressed in the "Enabling Functionality" section with a new parameter in `Pipeline`.

Related to https://github.com/scikit-learn/scikit-learn/pull/18444 - `get_feature_names_out` PR
Related to https://github.com/scikit-learn/scikit-learn/pull/16772 - `array_out` PR
Related to https://github.com/scikit-learn/scikit-learn/pull/18010 - `feature_names_out_` PR

CC @scikit-learn/core-devs

### CODE REVIEW

File: slep015/proposal.rst [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+``X``. This can be enabled by adding a ``array_input`` parameter to
+``Pipeline``::
+
+    pipe = make_pipeline(ct, MyTransformer(), LogisticRegression(),
+                         array_input='pandas')
+
+In this case, the pipeline will construct a pandas DataFrame to be inputted
+into ``MyTransformer`` and ``LogisticRegression``. The feature names
+will be constructed by calling ``get_feature_names_out`` as data is passed
+through the ``Pipeline``.
```
Comment (adrinjalali) [inline slep015/proposal.rst:93]: This implies it's the pipeline doing it. Or do you mean a pandas DF is passed around as your PR?

File: slep015/proposal.rst [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+Considerations
+##############
+
+This SLEP requires all estimators to store ``feature_names_in_` for all
```
Comment (adrinjalali) [inline slep015/proposal.rst:98]:
```suggestion
This SLEP requires all estimators to store ``feature_names_in_`` for all
```

File: slep015/proposal.rst [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+estimators, which will increase the size of the estimators. By default, a
+``Pipeline`` will only store ``feature_names_in_`` in the first step and
+the rest can be computed by slicing the pipeline at different steps. In other
+words, the additional space used will be at a minimal because only the
+input feature names from the first step are stored.
```
Comment (adrinjalali) [inline slep015/proposal.rst:103]: Then are we storing feature names in each step or not? This paragraph is a bit confusing to me.

Comment (adrinjalali) [reply slep015/proposal.rst:103]: Also, slicing from the middle of the pipeline, would lose the feature names.

**Review by adrinjalali — COMMENTED** [review]
I'm a bit confused as what your proposal really is. This sounds to me much closer to Andy's approach, but that doesn't pass the feature names during `fit` down the pipeline. Maybe you could expand on those bits a bit?

File: slep015/proposal.rst (hunk 3 above)
Comment (glemaitre) [reply slep015/proposal.rst:103]: Or we need to implement a `__getitem__` to recreate `feature_names_in_` when `start` of the slice is not `0`?

**Comment by thomasjpfan** [comment]
@adrinjalali I'm a bit confused as what your proposal really is. This sounds to me much closer to Andy's approach, but that doesn't pass the feature names during fit down the pipeline. Maybe you could expand on those bits a bit?

One of the reasons I wrote this SLEP was https://github.com/scikit-learn/scikit-learn/pull/16772#issuecomment-699960961. It highlighted the performance issues for sparse data which is related to the `InputArray` idea.

One can consider this SLEP a stepping stone toward passing feature names down the pipeline. Furthermore this SLEP is a prerequisite for the `array_out` PR:

1. `feature_names_in_` is required to make sure `fit` and `transform` has the same feature names.
2. `get_feature_names_out` needs to be implemented everywhere in the `array_out` PR, but privately.

This SLEP extends Andy's idea by including the interaction between `feature_names_in_` and `get_feature_names_out`. Since every estimator has the ability to store `feature_names_in_` users would not need to pass in the input names in `get_feature_names_out` in a pipeline or with a single estimator to get the output names.

Yes every step in a pipeline would not have the names, but these names can be obtained through slicing and if the estimator needs the name in `fit`, we can add a parameter in pipeline to tell it to construct the dataframe which it will pass to one of its steps. (This is an idea proposed in this SLEP but **not** a part of this SLEP)

**TLDR**: I think this SLEP is simpler and resolves some of the pain points we have with feature names.

File: slep015/proposal.rst (hunk 1 above)
Comment (thomasjpfan) [reply slep015/proposal.rst]: I am implying that Pipeline is doing this. I update the doc to make this point explicit. I also added more details on how this SLEP relates to ``array_out``.

File: slep015/proposal.rst (hunk 3 above)
Comment (thomasjpfan) [reply slep015/proposal.rst]: > Also, slicing from the middle of the pipeline, would lose the feature names.

We can make this work by constructing the ``feature_names_in_`` and attaching it to the new pipeline.

**Comment by thomasjpfan** [comment]
Updated SLEP is state that the ``Pipeline`` will have ``feature_names_in_``.

File: slep015/proposal.rst [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+be used for validation during non-``fit`` methods such as ``transform`` or
+``predict``. If the ``X`` is not a recognized container, then
+``feature_names_in_`` would be set to ``None``.
+
+Secondly, this SLEP proposes adding ``get_feature_names_out(input_names=None)``
+to all transformers. By default, the input features will be determined by the
+``feature_names_in_`` attribute. The feature names of a pipeline can then be
+easily extracted as follows::
+
+    pipe[:-1].get_feature_names_out()
```
Comment (ogrisel) [inline slep015/proposal.rst:66]:
```suggestion
    pipe[-1].get_feature_names_out()
```

Comment (adrinjalali) [reply slep015/proposal.rst:66]: the last step of the pipeline only has input feature names, not output, unless it's a transformer.

File: slep015/proposal.rst [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+   ``get_feature_names_out`` method is that it can be used when the input names
+   are not passed in ``fit``. This can happen in a pipeline when transformers
+   output ndarrays or sparse matrices without names. With a
+   ``feature_names_out_`` attribute, the estimator would require an array
+   container with feature names for ``feature_names_out_`` to be defined.
+
+3. [slep_012] : The ``InputArray`` was developed to work around the overhead
+    of using a pandas ``DataFrame`` or an xarray ``DataArray``. The
+    introduction of another data structure into the Python Data Ecosystem,
+    would be lead to more burden for third party estimator maintainers.
```
Comment (lorentzenchr) [inline slep015/proposal.rst:156]:
```suggestion
    would lead to more burden for third party estimator maintainers.
```
Or "would have lead"?

**Review by lorentzenchr — COMMENTED** [review]
@thomasjpfan Thanks for working on this! I've mainly nitpicks.

File: slep015/proposal.rst (hunk 4 above)
Comment (ogrisel) [reply slep015/proposal.rst:66]: ah right...

File: slep015/proposal.rst [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+2. [slep_007]_ : For ``SLEP007`` to function it still requires a mechanism to
+   pass feature names through a pipeline such as ``array_out``.
+   Furthermore, ``feature_names_out_`` would not be needed because it can be
+   computed with ``get_feature_names_out``. The benefit of the
+   ``get_feature_names_out`` method is that it can be used when the input names
+   are not passed in ``fit``. This can happen in a pipeline when transformers
+   output ndarrays or sparse matrices without names. With a
+   ``feature_names_out_`` attribute, the estimator would require an array
+   container with feature names for ``feature_names_out_`` to be defined.
```
[the remark this answers is not in the document]
Comment (thomasjpfan) [reply slep015/proposal.rst]: I was trying to say that for a `Pipeline` the feature names are not passed through each step and a `get_feature_names_out` method can be used to get the names with slicing.

I updated proposal with this clarification.

**Review by adrinjalali — APPROVED** [review]
I'm happy to have this merged. And looks good to me for a vote from my side :)

File: slep015/proposal.rst [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+                 .get_feature_names(orig_cat_cols))
+
+    feature_names = np.r_[cat_names, orig_num_cols]
+
+The ``feature_names`` extracted above corresponds to the features directly
+passed into ``LogisticRegression``. As demonstrated above, the process of
+extracting ``feature_names`` requires knowing the order of the selected
+categories in the ``ColumnTransformer``. Furthermore, if there is feature
+selection in the pipeline, such as ``SelectKBest``, the ``get_support`` method
+would need to be used to select column names that were selected.
```
Comment (jnothman) [inline slep015/proposal.rst:50]:
```suggestion
would need to be used to infer the column names that were selected.
```

File: slep015/proposal.rst [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+obtain the output feature names of the ``Pipeline`` itself.
+
+Enabling Functionality
+######################
+
+The following enhancements are **not** a part of this SLEP. These features are
+made possible if this SLEP gets accepted.
+
+1. As an alternative to slicing, we can add a
+   ``Pipeline.get_feature_names_in_at`` method to get the names at a specific
```
Comment (jnothman) [inline slep015/proposal.rst:82]: I find this name unpleasant, and don't see what's so much better than `Pipeline[-1].feature_names_in_`

File: slep015/proposal.rst [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+   step. This can be a simple alternative to slicing::
+
+      pipe.get_feature_names_in_at(-1)
+
+2. This SLEP enables us to implement an ``array_out`` keyword argument to
+   all ``transform`` methods to specify the array container outputted by
+   ``transform``. An implementation of ``array_out`` requires
+   ``feature_names_in_`` to validate that the names in ``fit`` and
+   ``transform`` are consistent. With the implementation of ``array_out`` needs
```
Comment (jnothman) [inline slep015/proposal.rst:91]:
```suggestion
   ``transform`` are consistent. An implementation of ``array_out`` needs
```

File: slep015/proposal.rst [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+selection in the pipeline, such as ``SelectKBest``, the ``get_support`` method
+would need to be used to select column names that were selected.
+
+Solution
+########
+
+This SLEP proposes adding the ``feature_names_in_`` attribute to all estimators
+that will extract the feature names of ``X`` during ``fit``. This will also
+be used for validation during non-``fit`` methods such as ``transform`` or
+``predict``. If the ``X`` is not a recognized container, then
```
Comment (jnothman) [inline slep015/proposal.rst:58]:
```suggestion
``predict``. If the ``X`` is not a recognized container with columns, then
```

File: slep015/proposal.rst [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+   through the ``Pipeline``. This feature implies that ``Pipeline`` is
+   doing the construction of the DataFrame.
+
+Considerations
+##############
+
+1. The ``get_feature_names_out`` will be constructed using the name generation
+   specification from :ref:`slep_007`.
+
+2. For a ``Pipeline`` with only one estimator, slicing will not work and one
```
Comment (jnothman) [inline slep015/proposal.rst:114]: I find this confusing. You're saying slicing will not work, but then showing an example with slicing? Or are you distinguishing slicing from indexing. What does slicing have to do with anything anyway??

**Review by jnothman — COMMENTED** [review]
I think this PR is good overall, but it needs to:
* specify the data type of `feature_names_in_` and the return type of `get_feature_names_out`: are they necessarily lists? any Sequence? are arrays permitted/required?
* address concerns about memory

If we are passing around a wide sparse matrix, and generating feature_names_in_ at each step in a pipeline, this could be consuming a lot of memory unnecessarily. Should we be introducing something like the following to avoid unnecessary memory usage with `self.feature_names_in_ = DefaultFeatureNames(X.shape[1])`?
```py
class DefaultFeatureNames(collections.abc.Sequence):
    def __init__(self, n_features):
        self.n_features = n_features

    def __len__(self):
        return self.n_features

    def __getitem__(self, sl):
        if isinstance(sl, slice):
            return [f"x{i}" for i in range(len(self))]
        elif isinstance(sl, tuple):
            raise NotImplementedError
        return f"x{sl}"

    def __iter__(self):
        return (f"x{i}" for i in range(len(self)))
```

File: slep015/proposal.rst (hunk 8 above)
Comment (thomasjpfan) [reply slep015/proposal.rst:82]: This will not work be default because the final step would not have the feature names if there is more than one step:

```py
pipe1 = make_pipeline(StandardScaler(), LogisticRegression())
# pipe1[-1].feature_names_in_ does not exist

pipe2 = make_pipeline(LogisticRegression())
# pipe2[-1].feature_names_in_ does exist
```

This proposal does not actually pass through the names at each step. Only the pipeline and the first step will have access to the input names.

I'll remove this point to make this SLEP shorter.

File: slep015/proposal.rst (hunk 11 above)
Comment (thomasjpfan) [reply slep015/proposal.rst:114]: I wanted to distinguish between the following two pipelines:

```py
pipe1 = make_pipeline(StandardScaler(), LogisticRegression())
pipe1[:-1].get_feature_names_out()  # this works

pipe2 = make_pipeline(LogisticRegression())
pipe2[:-1].get_feature_names_out()  # does not work
```

`pipe2[:-1]` fails because the slicing will produce a pipeline with no steps. Although, we can allow pipelines with no steps to get `pipe2[:-1].get_feature_names_out()` to work.

**Review by jnothman — COMMENTED** [review]
> This proposal does not actually pass through the names at each step. Only the pipeline and the first step will have access to the input names.

Oh! I don't think I understood this at all from the SLEP. Isn't that contradicted by "This SLEP proposes adding the ``feature_names_in_`` attribute to all estimators
that will extract the feature names of ``X`` during ``fit``."?

**Comment by thomasjpfan** [comment]
> This SLEP proposes adding the feature_names_in_ attribute to all estimators
that will extract the feature names of X during fit.

I was thinking of the "outer" most layer of the API. For non-meta estimators, `feature_names_in_` would be defined if the feature names can be extract from `X`.

For metaestimators, such as pipeline, they would have `pipe.feature_names_in_` defined but we would not guarantee that inner estimators would have `feature_names_in_` defined.

**Comment by jnothman** [comment]
I don't really get how you define "outer" or would ensure that from an
implementation perspective, unless either feature_names_in_ was being set
not by an estimator in its fit method, but by the caller of that
estimator's fit method; or, if feature_names_in_ was only set when X was in
a format that had names attached. Neither of these limitations are
discussed in the SLEP afaict.

**Comment by thomasjpfan** [comment]
> Neither of these limitations are discussed in the SLEP afaict.

Thank you for your thoughts, I'll update the SLEP accordingly.

> I don't really get how you define "outer"

I was using "outer" and "inner" in context of a metaestimator. The metaestimator would be the "outer" estimator, while all estimators inside the meta estimators are "inner" estimators.

> if feature_names_in_ was only set when X was in a format that had names attached.

This is the form I was considering.

> unless either feature_names_in_ was being set not by an estimator in its fit method, but by the caller of that estimator's fit method

An metaestimator could have all sorts of estimators that it uses internally. I think having the metaestimator be responsible for setting `feature_names_in_` for all its estimators would be too much of a requirement for metaestimator developers.

This SLEP is trying to proposal the bare minimum: If `X` has names and is passed to `fit`, then the estimator stores them as `feature_names_in_`.

Metaestimators is a case where I would want it to delegate this responsibly to its inner estimators. This way the metaestimator can construct `metaestimator.feature_names_in_` for itself. In the case of `Pipeline`, it would be the `feature_names_in_` of the first step. The issue is if the inner estimator does not set `feature_names_in_`, the metaestimator would have to set `metaestimator.feature_names_in_` to follow this SLEP.

**Comment by thomasjpfan** [comment]
> specify the data type of feature_names_in_ and the return type of get_feature_names_out: are they necessarily lists? any Sequence? are arrays permitted/required?

I was considering any `Sequence`: [`ffd0954` (#48)](https://github.com/scikit-learn/enhancement_proposals/pull/48/commits/ffd0954316e5cc6ec4cfe7b82045a8e09adc6feb).

@jnothman I have updated the SLEP to addresss your concerns:

1. Regarding meta-estimators: [`7537b15` (#48)](https://github.com/scikit-learn/enhancement_proposals/pull/48/commits/7537b15cb6c30d0b18e9075230e34b644830ee10)
2. Abstract now explicitly states that `feature_names_in_` are set in `fit`.
3. An example of slicing can fail with a pipeline with one step was added.

File: slep015/proposal.rst [hunk 12]
```diff
@@ -121,7 +121,10 @@ Considerations and Limitations
    a pipeline with no steps. We can work around this by allowing pipelines
    with no steps.
 
-3. Meta-estimators will delegate the setting and validation of
+3. ``feature_names_in_`` can be any 1-D ``Sequence``, such as an list or
```
Comment (jnothman) [inline slep015/proposal.rst:124]: But ndarray is not a sequence: https://github.com/numpy/numpy/issues/2776

Comment (thomasjpfan) [reply slep015/proposal.rst:124]: Maybe "Iterable that returns a string" would be enough.

In our discussions, I think we want to make sure the feature names are strings.

Comment (jnothman) [reply slep015/proposal.rst:124]: Hmm. We'd better accept Sequences and 1d array-likes whose elements are strings: pd.Index is not a sequence.

File: slep015/proposal.rst (hunk 11 above)
Comment (jnothman) [reply slep015/proposal.rst:111]: Okay, I agree this is a strange corner case, since it is the only way to construct a fitted empty Pipeline...

**Comment by jnothman** [comment]
So now my understanding (from conversation, not reviewing the SLEP) is that `feature_names_in_` should be set directly by every estimator that cannot delegate it.

Let's clarify:

* When input is a list of strings, `feature_names_in_ = None`.
* When input is a pd.DataFrame, `feature_names_in_ = X.columns`. Or is it `X.columns.copy()`?
* When input is a 2d array or sparse matrix, `feature_names_in_` is what?
* In the absence of a cross-library API for frame-like implementations, what happens when we don't support extracting column names from some data type in one version, but then add the capability in the next release. How does that affect other capabilities?
* How does a third party library know which data types are supported for storing column names? Do we give them a helper function to store/generate names in all supported cases?

**Comment by thomasjpfan** [comment]
> When input is a list of strings, feature_names_in_ = None.

If we want to be consistent with `n_feature_names` this would be the case. I think I would prefer it to not be defined, which means there is no validation in non-`fit` methods.

> When input is a pd.DataFrame, feature_names_in_ = X.columns. Or is it X.columns.copy()?

It would be safer to copy, but I would prefer not to.

> When input is a 2d array or sparse matrix, feature_names_in_ is what?

Not defined and will not validate.

> Do we provide ducktyped support for anything with .columns?
>
> In the absence of a cross-library API for frame-like implementations, what happens when we don't support extracting column names from some data type in one version, but then add the capability in the next release. How does that affect other behaviours?

I think the safest thing to do is to restrict this SLEP to pandas until we get a frame-like protocol. When this frame-like protocol is defined, then we can say we support frame-like objects.

> How does a third party library know which data types are supported for storing column names? Do we give them a helper function to store/generate names in all supported cases?

In https://github.com/scikit-learn/scikit-learn/pull/18010, I am pushing for using `_validate_data` to handle this. Internally, a third party estimator can also use `_check_feature_names`.

I think we should adjust the SLEP to make `feature_names_in_` optional. We can increase its adoption by demonstrating how useful it is with pipelines and other meta-estimators.

**Comment by jnothman** [comment]
But _validate_data isn't pubic. Since meta estimators need to handle the
case that the attribute is absent or None in any case, making this optional
seems reasonable.

File: slep015/proposal.rst (hunk 12 above)
Comment (jnothman) [reply slep015/proposal.rst:124]:
```suggestion
3. ``feature_names_in_`` can be any 1d array-like of strings, such as an list or
```

File: slep015/proposal.rst [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+      pipe2 = make_pipeline(LogisticRegression())
+      pipe[:-1].feature_names_in_  # Does not work
+
+   This is because `pipe2[:-1]` raises an error because it will result in
+   a pipeline with no steps. We can work around this by allowing pipelines
+   with no steps.
+
+3. ``feature_names_in_`` can be any 1-D ``Sequence``, such as an list or
+   an ndarray.
```
Comment (jnothman) [inline slep015/proposal.rst:125]: It might be worth noting that this allowance can avoid unnecessary memory consumption/copies, with reduced implementation complexity, although it may reduce usability a bit.

**Review by jnothman — COMMENTED** [review]
Regarding the discussion above, we need to make it clear that support for this is optional outside of the core library... Or we need a new discussion about how to make `_validate_data`'s capabilities publicly available (either by making it a public function in sklearn.utils, or by defining a "protected" estimator class API).

**Comment by lorentzenchr** [comment]
Is this PR superseded by acceptance of SLEP007 in #59?

**Comment by adrinjalali** [comment]
> Is this PR superseded by acceptance of SLEP007 in #59?

No, that one is about how the feature names are generated, this one is about how they're propagated in a pipeline.

**Comment by lorentzenchr** [comment]
Does this supersede #18, meaning it's an alternative?

**Comment by adrinjalali** [comment]
I'd say this is yet another effort to do the same thing as #18 wanted to do, but a lot more updated, after having gone down a few paths and not getting anywhere. I would need to read both again to say if this one is an alternative or just supersedes the other one.

**Comment by adrinjalali** [comment]
Also, I'd be happy to merge both of them and continue discussing on a separate issue.

**Comment by thomasjpfan** [comment]
A bunch of SLEP015 is already include in SLEP007 which has already been accepted. What SLEP015 adds would be the API for actually outputting pandas DataFrames.

I think https://github.com/scikit-learn/enhancement_proposals/pull/18 (SLEP008) has been superseded by SLEP007.

**Comment by adrinjalali** [comment]
back when we were writing sleps 7 and 8, I wrote 7 only to just talk about how we create the feature names, and 8 was more about the options we have to propagate them. Since then, I worked on a sklearn dataframe kinda object which we decided not to do, then worked on xarray, then Thomas worked on pandas, and the API also kinda evolved over time. I think we could at least focus on this SLEP for now, and then figure out what to do with other containers. We've gone back and forth with which container to use, or to use one at all, instead of propagating feature names alongside the data instead of with the data in a container.

**Comment by amueller** [comment]
This has been superseded by SLEP 18 #72 right?

**Comment by thomasjpfan** [comment]
I updated this PR so that this SLEP is now rejected.

**Comment by adrinjalali** [comment]
Thanks @thomasjpfan . Then I think we can merge.
